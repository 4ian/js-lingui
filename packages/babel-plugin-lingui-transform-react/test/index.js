import fs from 'fs'
import glob from 'glob'
import path from 'path'
import { transformFileSync, transform } from 'babel-core'

import plugin from '../src/index'

function getTestName(testPath) {
  return path.basename(testPath)
}

describe('babel-plugin-lingui-transform-react', function () {
  const babelOptions = {
    plugins: [
      'external-helpers',
      'syntax-jsx',
      'transform-remove-strict-mode',
      plugin
    ]
  }

  const transformCode = (code) => () => transform(code, babelOptions)

  glob.sync(path.join(__dirname, 'fixtures/*/')).forEach(testPath => {
    const testName = getTestName(testPath)
    const actualPath = path.relative(process.cwd(), path.join(testPath, 'actual.js'))
    const expectedPath = path.join(testPath, 'expected.js')

    it(testName, () => {
      const expected = fs.readFileSync(expectedPath, 'utf8')
      const actual = transformFileSync(actualPath, babelOptions).code.trim()

      expect(actual).toEqual(expected.trim())
    })
  })

  describe('validation', function () {
    describe('Plural/Select/SelectOrdinal', function () {
      it('value must be a variable', function () {
        const code = `<Plural value="42" one="Book" other="Books" />`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it('value is missing', function () {
        const code = `<Plural one="Book" other="Books" />`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it('offset must be number or string, not variable', function () {
        const code = `<Plural value={value} offset={offset} one="Book" other="Books" />`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it('plural forms are missing', function () {
        const plural = `<Plural value={value} />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `<Select value={value} />`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })

      it('plural forms missing fallback', function () {
        const plural = `<Plural value={value} one="Book" />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `<Select value={value} one="Book" />`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} one="Book" />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })

      it('plural rules must be valid', function () {
        const plural = `<Plural value={value} three="Invalid" one="Book" other="Books" />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} three="Invalid" one="st" other="rd" />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })
    })

    describe('Date/Number', function () {
      it('value must be a variable', function () {
        expect(transformCode('<NumberFormat />')).toThrowErrorMatchingSnapshot()
        expect(transformCode('<NumberFormat value="42" />')).toThrowErrorMatchingSnapshot()
      })

      it('format must be string, variable or object with custom format', function () {
        expect(transformCode('<NumberFormat value={value} format="custom" />')).not.toThrow()
        expect(transformCode('<NumberFormat value={value} format={"custom"} />')).not.toThrow()
        expect(transformCode('<NumberFormat value={value} format={custom} />')).not.toThrow()
        expect(transformCode('<NumberFormat value={value} format={{ digits: 4 }} />')).not.toThrow()
        expect(transformCode('<NumberFormat value={value} format={42} />')).toThrowErrorMatchingSnapshot()
      })

      it('value must be a variable', function () {
        expect(transformCode('<DateFormat />')).toThrowErrorMatchingSnapshot()
        expect(transformCode('<DateFormat value="42" />')).toThrowErrorMatchingSnapshot()
      })
    })
  })
})
