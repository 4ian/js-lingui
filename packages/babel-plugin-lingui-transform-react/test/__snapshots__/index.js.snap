// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-lingui-transform-react validation Date/Number value must be a variable 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Value argument is missing."`;

exports[`babel-plugin-lingui-transform-react validation Date/Number value must be a variable 2`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Value must be a variable."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal offset must be number or string, not variable 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Offset argument cannot be a variable."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms are missing 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing plural choices. At least fallback argument 'other' is required."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms are missing 2`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing select choices. At least fallback argument 'other' is required."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms are missing 3`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing selectordinal choices. At least fallback argument 'other' is required."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms missing fallback 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing fallback argument 'other'."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms missing fallback 2`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing fallback argument 'other'."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural forms missing fallback 3`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Missing fallback argument 'other'."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural rules must be valid 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English)."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal plural rules must be valid 2`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English)."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal value is missing 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Value argument is missing."`;

exports[`babel-plugin-lingui-transform-react validation Plural/Select/SelectOrdinal value must be a variable 1`] = `"packages/babel-plugin-lingui-transform-react/test/fixtures/nested-components/actual.js: Value must be a variable."`;
